src/coreComponents/mesh/MeshManager.cpp

```diff
@@ -75,13 +78,80 @@ void MeshManager::generateMeshLevels( DomainPartition & domain )
   {
-    meshGen.importFields( domain );
+    material.forWrappers( [&]( WrapperBase const & wrapper )
+    {
+      if( wrapper.sizedFromParent() )
+      {
+        materialWrapperNames.insert( ConstitutiveBase::makeFieldName( material.getName(), wrapper.getName() ) );
+      }
+    } );
   } );
+  return materialWrapperNames;
 }
 
+void MeshManager::importFields( DomainPartition & domain )
+{
+  GEOSX_MARK_FUNCTION;
+  forSubGroups< MeshGeneratorBase >( [&]( MeshGeneratorBase & generator )
+  {
+    if( !domain.hasMeshBody( generator.getName() ) )
+      return;
+
+    FieldIdentifiers fieldsToBeSync;
+    std::map< string, string > fieldNamesMapping = generator.getFieldsMapping();
+
+    ElementRegionManager & elemManager = domain.getMeshBody( generator.getName() ).getBaseDiscretization().getElemManager();
+    elemManager.forElementSubRegionsComplete< CellElementSubRegion >( [&]( localIndex,
+                                                                           localIndex,
+                                                                           ElementRegionBase const & region,
+                                                                           CellElementSubRegion & subRegion )
+    {
+      std::unordered_set< string > const materialWrapperNames = getMaterialWrapperNames( subRegion );
+      // Writing properties
+      for( auto const & pair : fieldNamesMapping )
+      {
+        string const & meshFieldName = pair.first;
+        string const & geosxFieldName = pair.second;
+        // Find destination
+        if( !subRegion.hasWrapper( geosxFieldName ) )
+        {
+          // Skip - the user may have not enabled a particular physics model/solver on this dstRegion.
+          GEOSX_LOG_LEVEL_RANK_0( 1, GEOSX_FMT( "Skipping import of {} -> {} on {}/{} (field not found)",
+                                                meshFieldName, geosxFieldName, region.getName(), subRegion.getName() ) );
+
+          continue;
+        }
+
+        // Now that we know that the subRegion has this wrapper, we can add the geosxFieldName to the list of fields to
+        // synchronize
+        fieldsToBeSync.addElementFields( {geosxFieldName}, {region.getName()} );
+        WrapperBase & wrapper = subRegion.getWrapperBase( geosxFieldName );
+        GEOSX_LOG_LEVEL_RANK_0( 1, GEOSX_FMT( "Importing field {} -> {} on {}/{}",
+                                              meshFieldName, geosxFieldName,
+                                              region.getName(), subRegion.getName() ) );
+
+        bool const isMaterialField = materialWrapperNames.count( geosxFieldName ) > 0 && wrapper.numArrayDims() > 1;
+        generator.importFieldsOnArray( region.getName(), meshFieldName, isMaterialField, wrapper );
+      }
+    } );
+
+    CommunicationTools::getInstance().synchronizeFields( fieldsToBeSync,
+                                                         domain.getMeshBody( generator.getName() ).getBaseDiscretization(),
```

Comment:
Maybe extract these calls as a separate variable, since it's already done line 114?